addons/sms/models/mail_thread.py

```diff
@@ -24,36 +24,201 @@ def _get_default_sms_recipients(self):
+                    break
+
+            if valid_number:
+                result[record.id] = {
+                    'partner': all_partners[0] if all_partners else self.env['res.partner'],
+                    'sanitized': valid_number, 'number': valid_number,
+                }
+            elif all_partners:
+                partner_number, partner = False, self.env['res.partner']
+                for partner in all_partners:
+                    partner_number = partner.mobile or partner.phone
+                    if partner_number:
+                        partner_number = phone_validation.phone_sanitize_numbers_string_w_record(partner_number, record)[partner_number]['sanitized']
+                    if partner_number:
+                        break
+
+                if partner_number:
+                    result[record.id] = {'partner': partner, 'sanitized': partner_number, 'number': partner_number}
+                else:
+                    result[record.id] = {'partner': partner, 'sanitized': False, 'number': partner.mobile or partner.phone}
+            elif all_numbers:
+                result[record.id] = {'partner': self.env['res.partner'], 'sanitized': False, 'number': all_numbers[0]}
+            else:
+                result[record.id] = {'partner': self.env['res.partner'], 'sanitized': False, 'number': False}
+        return result
+
+    def _message_sms(self, body, subtype_id=False, partner_ids=False, number_field=False, numbers=False):
+        """ Main method to post a message on a record using SMS-based notification
+        method.
+
+        :param body: content of SMS;
+        :param subtype_id: mail.message.subtype used in mail.message associated
+          to the sms notification process;
+        :param partner_ids: if set is a record set of partners to notify;
+        :param number_field: if set is a name of field to use on current record
+          to compute a number to notify;
+        :param numbers: if set is a list of phone numbers to notify;
+        """
+        self.ensure_one()
+
+        if number_field or (partner_ids is False and numbers is False):
+            info = self._sms_get_recipients_info(force_field=number_field)[self.id]
+            info_partner_ids = info['partner'].ids if info['partner'] else False
+            info_number = info['sanitized'] if info['sanitized'] else info['number']
+            if info_partner_ids and info_number:
+                pid_to_number = self.env.context.get('notify_sms_pid_to_number', {})
+                pid_to_number[info_partner_ids[0]] = info_number
+                self = self.with_context(notify_sms_pid_to_number=pid_to_number)
+            if info_partner_ids:
+                partner_ids = info_partner_ids + (partner_ids or [])
+            if info_number and not info_partner_ids:
+                numbers = [info_number] + (numbers or [])
+
+        if numbers:
+            self = self.with_context(notify_sms_numbers=numbers)
```

Comment:
As `message_post` allows `kwargs`, can't it be an explicit param ?

Reply:
Done